doc/1.4/language.md

```diff
@@ -3815,6 +3828,271 @@ independent method callback(int i, void *aux) {
+        log spec_viol: "%s was written to when not allowed", qname;
+    }
+
+    method write_register(uint64 written, uint64 enabled_bytes,
+                          void *aux) default {
+        if (write_allowed()) {
+            default(written, enabled_bytes, aux);
+        } else {
+            on_write_attempted_when_not_allowed();
+        }
+    }
+}
+
+template very_complex_gated_register is (very_complex_register,
+                                         gated_register) {
+    // No sensible way to combine the two implementations by calling both.
+    // Even if there were, calling both implementations would cause each field
+    // of the register to be written to multiple times, potentially duplicating
+    // side-effects, which is undesirable.
+    // Instead, very_complex_register is chosen as the base implementation
+    // called, and the behaviour of gated_register is replicated around that
+    // call.
+    method write_register(uint64 written, uint64 enabled_bytes,
+                          void *aux) default {
+        if (write_allowed()) {
+            this.templates.very_complex_register.write_register(
+                written, enabled_bytes, aux);
+        } else {
+            on_write_attempted_when_not_allowed();
+        }
+    }
+}
+
+in each (gated_register, very_complex_register) {
+    is very_complex_gated_register;
+}
+```
+
+A template-qualified method implementation call is resolved by using
+the method implementation provided to the object by the named template.
+If no such implementation is provided (whether it be because the template does
+not specify one, or specifies one which is not provided to the object due to its
+definition being eliminated by an [`#if`](#conditional-objects)), then the
+ancestor templates of the named template are recursively searched for the
+highest-ranking (most specific) implementation provided by them. If the ancestor
+templates provide multiple hierarchically unrelated implementations, then the
+choice is ambiguous and the call will be rejected by the compiler. In this case,
+the modeller must refine the template-qualified method implementation call to
+name the ancestor template whose implementation they would like to use.
+
+A template-qualified method implementation call done via [a value of template
+type](#templates-as-types) &mdash; most significantly, done via the (implicit)
+`this` within the body of a `shared` method &mdash; function differently in
+comparison to with compile-time object references. The specified template
+must be an ancestor template of the value's template type, the <tt>object</tt>
+template, or the template type itself; and the specified template **must
+provide or inherit a `shared` implementation of the named method**. It is not
+sufficient that the method is simply _declared_ `shared` such that it is part
+of the template type: the implementation itself must also be `shared`.
+For example, the following is permitted:
```

Comment:
This part of the TQMIC documentation goes on very long for a rather rare circumstance for anyone to need deal with. This is really just to highlight how unusual these restrictions are (that being that the method is part of the template type is _not_ enough), and how one would go about dealing with that in case you _do_ want to a non-shared implemented TQMIC from a value of template type. If I should cut this down, what should I cut away.

Reply:
One idea is to move details into the docstring of the corresponding error message(s) in messages.py, if that's applicable. Those are in an appendix, where lengthy details are less of a concern. People are unlikely to explore that appendix by themselves, so the main text should probably contain forward-links if they provide interesting info.

Reply:
Not too bad of an idea, other than it makes reading the source of `messages.py` a strange experience when one docstring is suddenly miles longer than the others.

Reply:
Not done yet, as it's not clearly a better solution.

Reply:
I think it's OK to make messages.py weird, I'm more concerned that it might make the refmanual look weird. I guess one way to find out is to try moving it and see if it does look outrageous.

And as a general direction, it would not disturb me if we'd move docs on error corners to messages.py.

Reply:
Alright, gonna do it then.

Reply:
Done. You'll have to run `./documentation` and see what the result looks like. In my opinion, it _does_ stick out quite a bit in the Messages section, _but_ it is acceptable.